src/Simulation/qdk_sim_rs/src/linalg/decompositions/lu.rs

```diff
@@ -0,0 +1,293 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+use std::any::type_name;
+
+use cauchy::Scalar;
+use ndarray::{Array1, ArrayBase, Axis, Data, Ix1, Ix2, OwnedRepr, RawData};
+use num_traits::FromPrimitive;
+
+use crate::{
+    error::QdkSimError,
+    linalg::array_ext::{RemoveAxisExt, ShapeExt},
+};
+
+/// Represents the output of an LU decomposition acting on a matrix.
+#[derive(Debug)]
+pub struct LU<A, S>
+where
+    S: Data<Elem = A>,
+    A: Scalar,
+{
+    pub(self) factors: ArrayBase<S, Ix2>,
+    pub(self) pivots: Array1<usize>,
+}
+
+/// Represents types that represent matrices that are decomposable into lower-
+/// and upper-triangular parts.
+pub trait LUDecomposable {
+    /// The element type for the output matrices.
```

Comment:
Are `LUDecompasable` and `LUDecomposition` also extension traits? I only see an impl for one type each. Is it needed for crate consumers to write their own impls?

Reply:
They're note extension traits, no, but an attempt to generalize 𝐿𝑈 decompositions to allow for application over other kinds of scalars and data structures including precomputed decompositions.

Reply:
If it's a generalization, that would imply there should be multiple impls for each trait, right? Are you planning to add some later?

Reply:
Not necessarily, but the corresponding split for eigendecompositions has proved useful (allows for precomputed decompositions as well, for example) such that leaving it open for consumers to write their own implementations may be useful as well. That does to some level presuppose using the linear algebra functionality outside of the context of the `qdk_sim` crate, but ideally I'd love to split that out into its own crate.

Reply:
Ok, sounds good, thanks for explaining.

Reply:
Absolutely, thanks for the comment! ♥